cpp/src/copying/contiguous_split.cu

```diff
@@ -14,465 +14,923 @@
-
-  // compute the rest of the column sizes (non-string columns, and total buffer size)
-  size_t total_size      = 0;
-  size_type column_index = 0;
-  std::for_each(
-    t.begin(), t.end(), [&total_size, &column_index, &split_info](cudf::column_view const& c) {
-      total_size +=
-        cudf::type_dispatcher(c.type(), column_buffer_size_functor{}, c, split_info[column_index]);
-      column_index++;
-    });
+struct split_key_functor {
+  int num_columns;
+  int operator() __device__(int buf_index) { return buf_index / num_columns; }
+};
 
-  // allocate
-  auto device_buf = std::make_unique<rmm::device_buffer>(total_size, stream, mr);
-  char* buf       = static_cast<char*>(device_buf->data());
-
-  // copy (this would be cleaner with a std::transform, but there's an nvcc compiler issue in the
-  // way)
-  std::vector<column_view> out_cols;
-  out_cols.reserve(t.num_columns());
-
-  column_index = 0;
-  std::for_each(
-    t.begin(), t.end(), [&out_cols, &buf, &column_index, &split_info](cudf::column_view const& c) {
-      cudf::type_dispatcher(
-        c.type(), column_copy_functor{}, c, split_info[column_index], buf, out_cols);
-      column_index++;
-    });
+/**
+ * @brief Output iterator for writing values to the dst_offset field of the
+ * dst_buf_info struct
+ */
+struct dst_offset_output_iterator {
+  dst_buf_info* c;
+  using value_type        = int;
+  using difference_type   = int;
+  using pointer           = int*;
+  using reference         = int&;
+  using iterator_category = thrust::output_device_iterator_tag;
+
+  dst_offset_output_iterator operator+ __host__ __device__(int i)
+  {
+    return dst_offset_output_iterator{c + i};
+  }
 
-  return contiguous_split_result{cudf::table_view{out_cols}, std::move(device_buf)};
-}
+  void operator++ __host__ __device__() { c++; }
+
+  reference operator[] __device__(int i) { return dereference(c + i); }
+  reference operator* __device__() { return dereference(c); }
+
+ private:
+  reference __device__ dereference(dst_buf_info* c) { return c->dst_offset; }
+};
+
+/**
+ * @brief Functor for computing size of data elements for a given cudf type.
+ *
+ * Note: columns types which themselves inherently have no data (strings, lists,
+ * structs) return 0.
+ *
+ */
+struct size_of_helper {
+  template <typename T>
+  constexpr std::enable_if_t<not is_fixed_width<T>(), int> __device__ operator()() const
+  {
+    return 0;
+  }
+
+  template <typename T>
+  constexpr std::enable_if_t<is_fixed_width<T>(), int> __device__ operator()() const noexcept
+  {
+    return sizeof(T);
+  }
+};
 
 };  // anonymous namespace
 
+namespace detail {
+
 std::vector<contiguous_split_result> contiguous_split(cudf::table_view const& input,
                                                       std::vector<size_type> const& splits,
                                                       rmm::mr::device_memory_resource* mr,
                                                       cudaStream_t stream)
```

Comment:
Depending on whether my PR is merged first. Either way this will be the norm going forward.

```suggestion
                                                      rmm::cuda_stream_view stream,
                                                      rmm::mr::device_memory_resource* mr)
```